modules/standard/Sets.chpl

```diff
@@ -0,0 +1,700 @@
+    iterator.
+
+    The set type is not parallel safe by default. For situations in which
+    such protections are desirable, parallel safety can be enabled by setting
+    `parSafe = true` in any set constructor. A set constructed from another
+    set inherits the parallel safety mode of that set by default.
+  */
+  record set {
+
+    /* The type of the elements contained in this set. */
+    type eltType;
+
+    /* If `true`, this set will perform parallel safe operations. */
+    param parSafe = false;
+
+    pragma "no doc"
+    var _lock$ = if parSafe then new _LockWrapper() else none;
+
+    pragma "no doc"
+    var _dom: domain(eltType, parSafe=false);
+
+    /*
+      Initializes an empty set containing elements of the given type.
+
+      :arg eltType: The type of the elements of this set.
+      :arg parSafe: If `true`, this set will use parallel safe operations.
+    */
+    proc init(type eltType, param parSafe=false) {
+      this.eltType = eltType;
+      this.parSafe = parSafe;
+    }
+
+    /*
+      Initialize this set with a unique copy of each element contained in
+      `iterable`. If an element from `iterable` is already contained in this
+      set, it will not be added again. The formal `iterable` must be a type
+      with an iterator defined for it.
+
+      :arg iterable: A collection of elements to add to this set.
+      :arg parSafe: If `true`, this set will use parallel safe operations.
+    */
+    proc init(type eltType, param parSafe, iterable) {
+      this.eltType = eltType;
+      this.parSafe = parSafe;
+      this.complete();
+      //
+      // TODO: Is there a reliable way to get iterable size, here? If so, we
+      // could make a call to `requestCapacity`.
+      //
+      for x in iterable do
+        _dom.add(x);
+    }
+
+    /*
+      Initialize this set with a copy of each of the elements contained in
+      the set `other`. This set will inherit the `parSafe` value of the
+      set `other`.
+
+      :arg other: A set to initialize this set with.
+    */
+    proc init=(const ref other: set(?t, ?)) {
+      this.eltType = t;
+      this.parSafe = other.parSafe;
+      this.complete();
+      for x in other {
+        //
+        // TODO: Not entirely sure if this inner copy is necessary, but we
+        // want to avoid potentially locking over and over.
+        //
+        var cpy = x;
+        _dom.add(cpy);
+      }
+    }
+
+    pragma "no doc"
+    inline proc deinit() {
+      clear();
+    }
+
+    pragma "no doc"
+    inline proc _enter() {
+      if parSafe then
+        _lock$.lock();
+    }
+
+    pragma "no doc"
+    inline proc _leave() {
+      if parSafe then
+        _lock$.unlock();
+    }
+ 
+    /*
+      Add a copy of the element `x` to this set. Does nothing if this set
+      already contains an element equal to the value of `x`.
+
+      :arg x: The element to add to this set.
+    */
+    proc add(in x: eltType) {
+      _enter();
```

Comment:
Should these methods use `on` statements? What if the `set` is remote?

Reply:
Does this mean that the code for every method should be encased in the following sort of on statement...?

```on this do ...```

And should the same be done for `list` methods as well?

Reply:
If you're acquiring a lock, you should always migrate to the locale the lock is allocated on, with very little exception.

Reply:
Thanks Louis!